src/base/dynamic_annotations.c

```diff
@@ -43,6 +43,19 @@
 #include "base/dynamic_annotations.h"
 #include "getenv_safe.h" // for TCMallocGetenvSafe
 
+static int running_on_valgrind_preload = -1;
+void __attribute__ ((constructor)) premain() {
```

Comment:
this will not work on non-gcc/non-clang compilers (and intel on gnu/linux also supports those).

Can you elaborate what is the issue you're fixing or improving ?

Reply:
> Can you elaborate what is the issue you're fixing or improving ?

Currently gperftools doesn't detect by itself that it runs under Valgrind. Detection is based on the compiler define value ```RUNNING_ON_VALGRIND``` which isn't set automatically, and alternatively on the environment variable ```RUNNING_ON_VALGRIND``` which can be set by the user. Both aren't automatic.

This patch allows gperftools to auto-detect Valgrind on Linix-like systems.

I maintain the gperftools port on FreeBSD. The user contacted me that he was in pain because there is no auto-detection of Valgrind in gperftools. I added this patch to the port, which satisfied the user.

This solution's idea is taken from [here](https://stackoverflow.com/questions/365458/how-can-i-detect-if-a-program-is-running-from-within-valgrind).

Reply:
I see. Yes, this seems justified. So just implementation needs to be different. First, existing code that runs (apparently insufficient) checks for valgrind is running early enough. No need to have extra "premain" thingy doing it.

Second, getenv sometimes works, but sometimes it is too early for it to work so early during first call to malloc. So we have special function just for that, inspecting environment early in program startup. There is GetenvBeforeMain.

Also maybe, but not sure, I'd consider to maybe only check LD_PRELOAD on unix-y platforms. Or maybe not worth any complexity to add those more ifdefs.